File: 1072-next-greater-node-in-linked-list/1072-next-greater-node-in-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<int> nextLargerNodes(ListNode* head) {
        std::vector<int> res;
        std::stack<int> st;
        while (head)
        {
            res.push_back(head->val);
            head = head->next;
        }
        for (int i = res.size() - 1; i >= 0; i--)
        {
            int temp = res[i];
            while (!st.empty() && st.top() <= temp)
            {
                st.pop();
            }
            if (st.empty())
            {
                res[i] = 0;
            }
            else
            {
                res[i] = st.top();
            }
            st.push(temp);
        }
        return res;
        // ListNode* mover = head;
        // std::vector<int> largerNodes;
        // bool notFound = true;
        // while (mover != NULL)
        // {
        //     ListNode* temp = mover->next;
        //     while (temp != NULL)
        //     {
        //         if (temp->val > mover->val)
        //         {
        //             notFound = false;
        //             largerNodes.push_back(temp->val);
        //             break;
        //         }
        //         temp = temp->next;
        //     }
        //     if (notFound)
        //     {
        //         largerNodes.push_back(0);
        //     }
        //     notFound = true;
        //     mover = mover->next;
        // }
        // return largerNodes;
    }
};

File: 1072-next-greater-node-in-linked-list/1072-next-greater-node-in-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int[] nextLargerNodes(ListNode head) {
        Stack<Integer> st = new Stack<>();
        ArrayList<Integer> res = new ArrayList<>();
        while (head != null)
        {
            res.add(head.val);
            head = head.next;
        }
        int n = res.size();
        for (int i = n - 1; i >= 0; i--)
        {
            int temp = res.get(i);
            while (!st.empty() && st.peek() <= temp)
            {
                st.pop();
            }
            if (st.empty())
            {
                res.set(i, 0);
            }
            else
            {
                res.set(i, st.peek());
            }
            st.push(temp);
        }
        int[] fin = new int[n];
        for (int i = 0; i < n; i++)
        {
            fin[i] = res.get(i);
        }
        return fin;
    }
}